flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java

```diff
@@ -51,39 +48,52 @@
  * @param <TXN> Transaction to store all of the information required to handle a transaction (must be Serializable)
  */
 @PublicEvolving
-public abstract class TwoPhaseCommitSinkFunction<IN, TXN extends Serializable>
+public abstract class TwoPhaseCommitSinkFunction<IN, TXN>
 		extends RichSinkFunction<IN>
 		implements CheckpointedFunction, CheckpointListener {
 
 	private static final Logger LOG = LoggerFactory.getLogger(TwoPhaseCommitSinkFunction.class);
 
-	protected final ListStateDescriptor<TransactionAndCheckpoint<TXN>> pendingCommitTransactionsDescriptor;
+	protected final ListStateDescriptor<Map<Long, TXN>> pendingCommitTransactionsDescriptor;
 	protected final ListStateDescriptor<TXN> pendingTransactionsDescriptor;
 
-	protected final List<TransactionAndCheckpoint<TXN>> pendingCommitTransactions = new ArrayList<>();
+	protected final LinkedHashMap<Long, TXN> pendingCommitTransactions = new LinkedHashMap<>();
 
 	@Nullable
 	protected TXN currentTransaction;
 	protected ListState<TXN> pendingTransactionsState;
-	protected ListState<TransactionAndCheckpoint<TXN>> pendingCommitTransactionsState;
-
-	public TwoPhaseCommitSinkFunction(Class<TXN> txnClass) {
-		this(
-			TypeInformation.of(txnClass),
-			TypeInformation.of(new TypeHint<TransactionAndCheckpoint<TXN>>() {}));
-	}
+	protected ListState<Map<Long, TXN>> pendingCommitTransactionsState;
```

Comment:
I think this has to be `ListState<Tuple2<Long, TXN>>` or the original `ListState<TransactionAndCheckpoint<TXN>>`.

Using a map instead of a list for `pendingCommitTransactions ` is ok for bookkeeping, but when snapshotting this map we need to make sure the snapshotted transactions are state elements that can be redistributable independent of each other.